src/mantaray-v1.ts

```diff
@@ -0,0 +1,738 @@
+    }
+
+    this._obfuscationKey = obfuscationKey
+    this.makeDirty()
+  }
+
+  public get obfuscationKey(): Bytes<32> {
+    return this._obfuscationKey
+  }
+
+  public set nodeMetadata(metadata: MetadataMapping | undefined) {
+    this._nodeMetadata = metadata
+    this.makeDirty()
+  }
+
+  public get nodeMetadata(): MetadataMapping | undefined {
+    return this._nodeMetadata
+  }
+
+  public set forkMetadata(metadata: MetadataMapping | undefined) {
+    this._forkMetadata = metadata
+    this.makeDirty()
+  }
+
+  public get forkMetadata(): MetadataMapping | undefined {
+    return this._forkMetadata
+  }
+
+  public set forkMetadataSegmentSize(value: number) {
+    if (value > 31) throw new Error(`forkMetadataSegmentSize is greater than 31. Got: ${value}`)
+    this._forkMetadataSegmentSize = value
+  }
+
+  public get forkMetadataSegmentSize(): number {
+    return this._forkMetadataSegmentSize
+  }
+
+  public get isEdge(): boolean {
+    return this._isEdge
+  }
+
+  public get hasEntry(): boolean {
+    return this._hasEntry
+  }
+
+  public get encEntry(): boolean {
+    return this._encEntry
+  }
+
+  public get metadata(): MetadataMapping | undefined {
+    if (!this._forkMetadata && !this._nodeMetadata) return undefined
+
+    return {
+      ...this.nodeMetadata,
+      ...this._forkMetadata,
+    }
+  }
+
+  /// Node type related functions
+  /// dirty flag is not necessary to be set
+
+  public isValueType(): boolean {
+    return this._hasEntry
+  }
+
+  /**
+   * The node either has metadata on node level or fork level
+   * for forkMetadata parent node has to be fetched
+   */
+  public isWithMetadataType(): boolean {
+    return Boolean(this._forkMetadata) || Boolean(this._nodeMetadata)
+  }
+
+  /// BL methods
+
+  public addFork(
+    path: Uint8Array,
+    attributes?: {
+      entry?: Reference
+      nodeMetadata?: MetadataMapping
+      forkMetadata?: MetadataMapping
+      obfuscationKeyGenerator?: Random32BytesFn
+    },
+  ): void {
+    const entry: Reference | undefined = attributes?.entry
+    const nodeMetadata: MetadataMapping | undefined = attributes?.nodeMetadata
+    const forkMetadata: MetadataMapping | undefined = attributes?.forkMetadata
+    const obfuscationKeyGenerator: Random32BytesFn | undefined = attributes?.obfuscationKeyGenerator
+
+    if (path.length === 0) {
+      if (entry) this.entry = entry
+
+      this.nodeMetadata = nodeMetadata
+      this.forkMetadata = forkMetadata
+
+      return
+    }
+
+    if (this.isDirty() && !this.forks) this.forks = {}
+
+    if (!this.forks) throw Error(`Fork mapping is not defined in the manifest`)
+
+    const fork = this.forks[path[0]]
+
+    if (!fork) {
+      const newNode = new MantarayNode()
+
+      if (!equalBytes(this._obfuscationKey, null32Bytes)) {
+        if (!obfuscationKeyGenerator) {
+          throw new RandomBytesFnUndefined()
+        }
+        newNode.obfuscationKey = obfuscationKeyGenerator()
+      }
```

Comment:
This is for example candidate for refactoring out to a separate method as it is even repeated several times.

Reply:
I think it is a bit overengineering the problem, there are only 2 occasions it appears and the share the same error class anyway.